### PR TITLE
Feature/operation service

### PR DESCRIPTION
**Pull request by dhelmert**
### Reason for this PR
- Multi user signing is a multi step process, requiring a workflow that allows approving operations at a later date.

### Changes in this PR
- Created the operation service, for handling operations. operations are wrappers, containing a transactions and information regarding who has approved, and still needs to approve in order for the transaction to be allowed to be executed

- [X] Test written
- [ ] Changelog updated

### CODE REVIEW

File: product_code/group_service/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# [v0.15.1](https://github.com/upb-uc4/University-Credits-4.0/compare/group-v0.15.1...group-v0.15.1) (2020-XX-XX)
```
Comment (lrschaef) [inline product_code/group_service/CHANGELOG.md:1]:
```suggestion
# [v0.16.1](https://github.com/upb-uc4/University-Credits-4.0/compare/group-v0.16.1...group-v0.16.1) (2021-XX-XX)
```

File: product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+  "OperationService service" should {
+
+    "fetch the hyperledger versions" in {
+      client.getHlfVersions.invoke().map { answer =>
+        answer shouldBe a[JsonHyperledgerVersion]
+      }
+    }
+
+    "get an operation as initiator" in {
```
Comment (lrschaef) [inline product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala:206]:
```suggestion
    "get an operation as its initiator" in {
```

File: product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    "get an operation as not the initiator" in {
+      prepare(Seq(operation1))
+      client.getOperation(operation1.operationId).handleRequestHeader(addAuthorizationHeader(student1)).invoke().map {
+        operation => operation should ===(operation1)
+      }
+    }
+
+    "get an operation with OwnerMismatch" in {
```
Comment (lrschaef) [inline product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala:220]:
```suggestion
    "fail with OwnerMismatch on trying to get an operation as an uninvolved user" in {
```

File: product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    "get operations where the user needs to do actions" in {
+      prepare(Seq(operation1, operation2))
+      client.getOperations(None, Some(true), None, None).handleRequestHeader(addAuthorizationHeader(admin0)).invoke().map {
+        operations => operations should contain theSameElementsAs Seq(operation2)
+      }
+    }
+
+    "get operations where the user initiated the operation" in {
```
Comment (lrschaef) [inline product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala:241]:
```suggestion
    "get all operations a user initiated, as the initiator" in {
```

File: product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+      prepare(Seq(operation1, operation2))
+      client.getOperations(Some(true), None, None, None).handleRequestHeader(addAuthorizationHeader(student0)).invoke().map {
+        operations => operations should contain theSameElementsAs Seq(operation1)
+      }
+    }
+
+    "get operations where the user asks for a state" in {
```
Comment (lrschaef) [inline product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala:248]:
```suggestion
    "get all operations with a specific state" in {
```

File: product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+      prepare(Seq(operation1, operation2, operation3))
+      client.getOperations(None, None, Some(OperationDataState.REJECTED.toString), None).handleRequestHeader(addAuthorizationHeader(student1)).invoke().map {
+        operations => operations should contain theSameElementsAs Seq(operation3)
+      }
+    }
+
+    "get operations where the user asks for his watchlist" in {
```
Comment (lrschaef) [inline product_code/operation_service/impl/src/test/scala/de/upb/cs/uc4/operation/impl/OperationServiceSpec.scala:255]:
```suggestion
    "get all operations from a user's watchlist, as the user" in {
```

**Review by lrschaef — CHANGES_REQUESTED** [review]
Changelogs need to be updated (Group Service and Operation Service)

**Review by jonas-thiele — APPROVED** [review]
Looks good to me

File: product_code/group_service/impl/src/main/resources/hyperledger_assets/connection_profile_kubernetes_local.yaml [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+#        LmNvbTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABGrsQ6oJpk6hDWf63HU3OSNd
+#        bou9KNw/VIee1IngPDI4YJU7O+Xa/XLJuwnFv7BpR8Ytl3f+njC8i/RZP2/svO+j
+#        XzBdMA4GA1UdDwEB/wQEAwIBpjAPBgNVHSUECDAGBgRVHSUAMA8GA1UdEwEB/wQF
+#        MAMBAf8wKQYDVR0OBCIEIIpzkSIZzxBWVIV5unlgZJuyu2XPEeP8+y1uB6LLA5Qr
+#        MAoGCCqGSM49BAMCA0gAMEUCIQDUh/+CC2dAICnYtACXspwUaaEbiyZxYIx+XDvW
+#        o8VVcgIgGz5S4iC5+xkxgeaISPfxKTTVy6yzTdYGzCw1vPppjzo=
+#        -----END CERTIFICATE-----
+#
+#    httpOptions:
+#      verify: false
```
[the remark this answers is not in the document]
Comment (david-buderus) [reply product_code/group_service/impl/src/main/resources/hyperledger_assets/connection_profile_kubernetes_local.yaml:89]: It is only used in local testing, so yeah it should stay

**Review by ldklenner — APPROVED** [review]

**Review by lrschaef — APPROVED** [review]

**Pull request merged**